### PR TITLE
[feat](nereids)add more rules to eliminate empty relation

### PR DESCRIPTION
**Pull request by englefly**
## Proposed changes
eliminate empty relations for following patterns:
topn->empty
sort->empty
distribute->empty
project->empty

Issue Number: close #xxx

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by englefly** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateEmptyRelation.java [hunk 1]
```diff
@@ -51,11 +51,18 @@ public List<Rule> buildRules() {
         return ImmutableList.of(
             // join->empty
             logicalJoin(any(), any())
-                .when(this::hasEmptyRelationChild)
-                .when(this::canReplaceJoinByEmptyRelation)
-                .then(join -> new LogicalEmptyRelation(
-                    ConnectContext.get().getStatementContext().getNextRelationId(),
-                    join.getOutput()))
+                // .when(this::hasEmptyRelationChild)
+                // .when(this::canReplaceJoinByEmptyRelation)
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateEmptyRelation.java:55]: remove useless code

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateEmptyRelation.java [hunk 2]
```diff
@@ -51,11 +51,18 @@ public List<Rule> buildRules() {
...
+                .then(join -> {
+                            if (hasEmptyRelationChild(join) && canReplaceJoinByEmptyRelation(join)
+                                    || bothChildrenEmpty(join)) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateEmptyRelation.java:58]: use when for better performance

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateEmptyRelation.java [hunk 3]
```diff
@@ -68,7 +75,13 @@ public List<Rule> buildRules() {
                     ConnectContext.get().getStatementContext().getNextRelationId(),
                     agg.getOutput())
                 ).toRule(RuleType.ELIMINATE_AGG_ON_EMPTYRELATION),
-
+            // proj->empty
+            logicalProject(logicalEmptyRelation())
+                    .then(project ->
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateEmptyRelation.java:80]: use `thenApply(ctx -> {ctx.statementContext})` for better performance

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateEmptyRelation.java (hunk 1 above)
Comment (englefly) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateEmptyRelation.java:55]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateEmptyRelation.java (hunk 2 above)
Comment (englefly) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateEmptyRelation.java]: done

**Comment by englefly** [comment]
run buildall

**Review by morrySnow — APPROVED** [review]

**Review by keanji-x — APPROVED** [review]

**Pull request merged**